feat: add quoting support to to_dict (#1052)

clumsy opened the pull request:
See #1052

Test plan:
[x] all existing test cases pass

[x] add new test case

[x] test locally:

Given:
```
[component:dist.ddp]
script=test.py
env=
	NCCL_IB_HCA='=mlx5_0:1,mlx5_1:1';
```

Observe:
```
test/0 [7]:environ({...'NCCL_IB_HCA': '=mlx5_0:1,mlx5_1:1', ...})
```

clumsy commented:
Please consider this PR, @kiukchung, @d4l3k, @tonykao8080 and sorry for using regex 🙂

clumsy commented:
No particular preference besides being convenient to see the exact value without unescaping, @andywag - would be happy with any working solution.

Do you think an alternative mechanism is more desirable? Do you mean escaping like `\=`?

andywag commented:
Yeah, I prefer the more standard escape like you mention but am fine with this since it's working.

andywag reviewed (approved):
lgtm

clumsy commented:
Addressed your comment in https://github.com/pytorch/torchx/pull/1049#issuecomment-2833817150 @andywag 

Please let me know if you prefer escaping after all, I have no preference but want to follow the principle of least astonishment.

andywag commented:
@clumsy Can you add the type to quoted_string quoted_values: List[str] = []? to fix a pyre issue

clumsy commented:
Done, @andywag
